### PR TITLE
 kubelet dra: lock before getting claimInfo CDIDevices and annotations fields

### PR DESCRIPTION
**Pull request by moshe010**
Currently claimInfo CDIDevices and annotations access directly without RLock. 
This can lead to concurrent read write error.
To avoid it we added RLock all before getting the CDIDevices and annotations.

#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by moshe010 — COMMENTED** [review]
/assign @bart0sh @klueska

File: pkg/kubelet/cm/container_manager_linux.go [hunk 1]
```diff
@@ -982,7 +982,8 @@ func (cm *containerManagerImpl) GetDynamicResources(pod *v1.Pod, container *v1.C
 		// a set of CDIDevices from a different kubelet plugin. In the future we may want to
 		// include the name of the kubelet plugin and/or other types of resources that are
 		// not CDIDevices (assuming the DRAmanager supports this).
-		for _, klPluginCdiDevices := range containerClaimInfo.CDIDevices {
+		claimInfoCdiDevices := containerClaimInfo.CloneCDIDevices()
```
[the remark this answers is not in the document]
Comment (moshe010) [reply pkg/kubelet/cm/container_manager_linux.go:985]: not sure if that good reason but clone is done by other manager as well [1]. So I wanted the code to be consistent with the other managers. 
Do you have  a compelling reason to prefer lock over clone? 

[1] - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/state/state_mem.go#L43-L56

File: pkg/kubelet/cm/dra/claiminfo.go [hunk 2]
```diff
@@ -62,6 +59,9 @@ func (info *ClaimInfo) addCDIDevices(pluginName string, cdiDevices []string) err
 		return fmt.Errorf("failed to generate container annotations, err: %+v", err)
 	}
 
+	info.Lock()
+	defer info.Unlock()
+
```
[the remark this answers is not in the document]
Comment (bart0sh) [reply pkg/kubelet/cm/dra/claiminfo.go:64]: yes, please undo. It was a wrong suggestion, sorry.

File: pkg/kubelet/cm/container_manager_linux.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bart0sh) [reply pkg/kubelet/cm/container_manager_linux.go:985]: Those examples are different. First `cpuset.CPUSet` doesn't have lock and second, in both cases cloned object is returned from the function, i.e. locking would be hard to do there.

[the remark this answers is not in the document]
Comment (moshe010) [reply pkg/kubelet/cm/container_manager_linux.go:985]: sorry I had urgent issues to take care and also currently I am OOO and will be back next week. 
Just to understand you suggestion is to expose the mutux from the ClaimInfo struct so we can lock it in  the right places?

[the remark this answers is not in the document]
Comment (bart0sh) [reply pkg/kubelet/cm/container_manager_linux.go:985]: @moshe010 yes, that's what I'd suggest to do.

File: pkg/kubelet/cm/dra/manager.go [hunk 3]
```diff
@@ -282,8 +284,8 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 				resourceHandle.Data)
 			if err != nil {
 				return fmt.Errorf(
-					"NodeUnprepareResource failed, pod: %s, claim UID: %s, claim name: %s, CDI devices: %s, err: %+v",
-					pod.Name, claimInfo.ClaimUID, claimInfo.ClaimName, claimInfo.CDIDevices, err)
+					"NodeUnprepareResource failed, pod: %s, claim UID: %s, claim name: %s, resource handle: %s, err: %+v",
+					pod.Name, claimInfo.ClaimUID, claimInfo.ClaimName, resourceHandle.Data, err)
```
Comment (bart0sh) [inline pkg/kubelet/cm/dra/manager.go:288]: `claimInfo` properties are referenced here and above. Why don't we lock it?

Comment (moshe010) [reply pkg/kubelet/cm/dra/manager.go:288]: The claimInfo is created in [1] and during the lifetime of the PrepareResources/GetResources/UnprepareResources we write only for following fields: PodUIDs (deletePodReference) , CDIDevices (addCDIDevices)  and annotations( addCDIDevices). 
So In my opinion we should only guard them. 

[1] - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/manager.go#L112-L119

**Comment by klueska** [comment]
Looks like there is an (unrelated) unit test failing. Triggering a retest...
/retest

**Pull request merged**